Script to frequently examine the database and reset persistence cache

sheridancbio opened the pull request:
- new script introduced
- property parsing functionality removed from dmp-import-vars-functions.sh and expanded

See comments in clear-persistence-caches-for-changed-studies.sh script for overview/summary of functionality

n1zea144 commented on import-scripts/clear-persistence-caches-for-changed-studies.sh:
```diff
@@ -0,0 +1,450 @@
+    new_study_info_from_db_filepath="$TMP_DIR_ROOT/${db}_study_info_new"
+    if ! [ -f "$previous_study_info_from_db_filepath" ] ; then
+        # if we have lost the previous study info, we cannot tell what has been modified. Prime the study info and trigger a full cache reset
+        write_current_study_list_from_db "$db" "$previous_study_info_from_db_filepath"
```
Check return status here.

sheridancbio replied:
Done. If the mysql command fails, a log message is generated and this function sees the failure status and returns failure too.

n1zea144 commented on import-scripts/clear-persistence-caches-for-changed-studies.sh:
```diff
@@ -0,0 +1,450 @@
+        clear_completed_triggers "$portal_db"
+        if ! database_is_triggered_as_needing_reset "$portal_db" ; then
+            if any_trigger_has_exceeded_maximum_wait "$portal_db"; then
+                trigger_db_as_needing_reset "$portal_db"
```
continue to the next loop iteration here.

sheridancbio replied:
Done

n1zea144 commented on import-scripts/clear-persistence-caches-for-changed-studies.sh:
```diff
@@ -0,0 +1,450 @@
+# usage : perform_selected_cache_reset_task
+function perform_selected_cache_reset_task() {
+    # if (unlikely) task has been successfully performed while we selected it, exit
+    if trigger_filepath_is_marked_as_succeeded "$selected_cache_reset_work_task_db" "$selected_cache_reset_work_task_trigger_filepath" ; then
```
Is this really needed?

sheridancbio replied:
Not really. The original thinking was that multiple processes can be running in PHASE TWO and it is hypothetically conceivable that one process might stop at the time between selecting the triggered task to perform and the obtaining of the lock to do the task. However, this case does not need to be covered because:
- after the trigger file is selected, the next step is to lock the file, making an interruption very unlikely
- it is also extremely unlikely that during this momentary vulnerability, another process could acquire the lock, accomplish the entire task, mark successful status and release the lock before this process advanced at all. It is practically certain that the other process would be holding the lock at the time we attempted to get it, so we would fail to obtain the lock and thereby exit.
- even if the envisioned scenario managed to happen, the only effect would be to clear the cache a second time for the trigger, which is not harmful.
Function and function call removed.

n1zea144 commented on import-scripts/clear-persistence-caches-for-changed-studies.sh:
```diff
@@ -0,0 +1,450 @@
+    study_id="${trigger_filename%.*}"
+    cache_api_key="${portal_db_to_cache_api_key[$selected_cache_reset_work_task_db]}"
+    cache_api_base_uri="${PORTAL_DB_TO_PORTAL_API_BASE_URI[$selected_cache_reset_work_task_db]}"
+    curl -X DELETE -H "X-API-KEY:$cache_api_key" -v -i "$cache_api_baseuri/" "$cache_api_base_uri/cache/$study_id"
```
Double check why two baseuri's are here.

sheridancbio replied:
that was an error. Thanks. That was fixed, and I removed the "-v" (verbose) option and the "-i" (include response headers) option to help reduce logging noise. I've also sent the standard output to /dev/null.

sheridancbio replied:
Actually, I tested and curl returns status 0 (success) even when the endpoint returns a failure message. Instead of redirecting to null, I pipe to grep to see if a "success" word is present in output. The word "Flushed" will appear in the output on success, so I test for it.

sheridancbio commented:
Added example/explanation to the docs with example trigger/status filename references.

sheridancbio commented:
I have also added an entry in mycrontab, and removed the call to the full recache at the end of the public import run script. With this running, that call will be redundant.
